eclair-core/src/main/scala/fr/acinq/eclair/wire/TlvCodecs.scala

```diff
@@ -28,52 +29,80 @@ import scodec.{Attempt, Codec, Err}
 
 object TlvCodecs {
 
-  private def variableSizeUInt64(size: Int, min: UInt64): Codec[UInt64] = minimalvalue(bytes(size).xmap(UInt64(_), _.toByteVector.takeRight(size)), min)
-
   /**
-   * Length-prefixed truncated uint64 (1 to 9 bytes unsigned integer).
+   * Truncated uint64 (0 to 8 bytes unsigned integer).
+   * The encoder minimally-encodes every value, and the decoder verifies that values are minimally-encoded.
+   * Note that this codec can only be used at the very end of a TLV record.
    */
-  val tu64: Codec[UInt64] = discriminated[UInt64].by(uint8)
-    .\(0x00) { case i if i < 0x01 => i }(variableSizeUInt64(0, 0x00))
-    .\(0x01) { case i if i < 0x0100 => i }(variableSizeUInt64(1, 0x01))
-    .\(0x02) { case i if i < 0x010000 => i }(variableSizeUInt64(2, 0x0100))
-    .\(0x03) { case i if i < 0x01000000 => i }(variableSizeUInt64(3, 0x010000))
-    .\(0x04) { case i if i < 0x0100000000L => i }(variableSizeUInt64(4, 0x01000000))
-    .\(0x05) { case i if i < 0x010000000000L => i }(variableSizeUInt64(5, 0x0100000000L))
-    .\(0x06) { case i if i < 0x01000000000000L => i }(variableSizeUInt64(6, 0x010000000000L))
-    .\(0x07) { case i if i < 0x0100000000000000L => i }(variableSizeUInt64(7, 0x01000000000000L))
-    .\(0x08) { case i if i <= UInt64.MaxValue => i }(variableSizeUInt64(8, 0x0100000000000000L))
+  val tu64: Codec[UInt64] = Codec(
```

Comment:
I have this, but it's not much better:

```scala
val tu64: Codec[UInt64] = Codec(
    (u: UInt64) => {
      def encode(bits: Int) = u.toByteVector.takeRight(bits)
      val b = u match {
        case u if u < 0x01 => encode(0)
        case u if u < 0x0100 => encode(1)
        case u if u < 0x010000 => encode(2)
        case u if u < 0x01000000 => encode(3)
        case u if u < 0x0100000000L => encode(4)
        case u if u < 0x010000000000L => encode(5)
        case u if u < 0x01000000000000L => encode(6)
        case u if u < 0x0100000000000000L => encode(7)
        case u if u <= UInt64.MaxValue => encode(8)
      }
      Attempt.successful(b.bits)
    },
    b => {
      def decode(min: UInt64) = minimalvalue(uint64, min).decode(b.padLeft(64))
      b.length match {
        case l if l <= 0 => decode(UInt64(0x00))
        case l if l <= 8 => decode(UInt64(0x01))
        case l if l <= 16 => decode(UInt64(0x0100))
        case l if l <= 24 => decode(UInt64(0x010000))
        case l if l <= 32 => decode(UInt64(0x01000000))
        case l if l <= 40 => decode(UInt64(0x0100000000L))
        case l if l <= 48 => decode(UInt64(0x010000000000L))
        case l if l <= 56 => decode(UInt64(0x01000000000000L))
        case l if l <= 64 => decode(UInt64(0x0100000000000000L))
        case _ => Attempt.failure(Err(s"too many bytes to decode for truncated uint64 (${b.toHex})"))
      }
    })
```

Reply:
I'll at least apply these refactorings, it looks a bit nicer but I don't see a fundamental way of escaping this overall pattern...at some point we need to have conditional branches for each of the integer slice range to ensure minimal encoding :/